<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>RNA velocity with scVelo — scvelo • velociraptor</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="RNA velocity with scVelo — scvelo"><meta property="og:description" content="Perform RNA velocity calculations with the scVelo package."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">velociraptor</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.15.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/velociraptor.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="divider">
    <li>
      <a href="../articles/index.html">More articles...</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/kevinrue/velociraptor/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>RNA velocity with <span class="pkg">scVelo</span></h1>
    <small class="dont-index">Source: <a href="https://github.com/kevinrue/velociraptor/blob/devel/R/scvelo.R" class="external-link"><code>R/scvelo.R</code></a></small>
    <div class="hidden name"><code>scvelo.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Perform RNA velocity calculations with the <span class="pkg">scVelo</span> package.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">scvelo</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'ANY'</span></span>
<span><span class="fu">scvelo</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  subset.row <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sf.X <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sf.spliced <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sf.unspliced <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  use.theirs <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  mode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"steady_state"</span>, <span class="st">"deterministic"</span>, <span class="st">"stochastic"</span>, <span class="st">"dynamical"</span><span class="op">)</span>,</span>
<span>  scvelo.params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  dimred <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncomponents <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  BSPARAM <span class="op">=</span> <span class="fu">bsparam</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SummarizedExperiment'</span></span>
<span><span class="fu">scvelo</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  assay.X <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  assay.spliced <span class="op">=</span> <span class="st">"spliced"</span>,</span>
<span>  assay.unspliced <span class="op">=</span> <span class="st">"unspliced"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SingleCellExperiment'</span></span>
<span><span class="fu">scvelo</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span>, sf.X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">sizeFactors</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, dimred <span class="op">=</span> <span class="cn">NULL</span>, use.dimred <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A named list of three matrices of the same dimensions where genes are in rows and cells are in columns.
The list should contain <code>"spliced"</code> and <code>"unspliced"</code> entries containing spliced and unspliced counts, respectively.
It should also contain an <code>"X"</code> entry containing the “usual” count matrix, see details below.</p>
<p>Alternatively, a SummarizedExperiment object containing three such matrices among its assays.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>For the generic, further arguments to pass to specific methods.
For the SummarizedExperiment and SingleCellExperiment methods, further arguments to pass to the ANY method.</p></dd>


<dt id="arg-subset-row">subset.row<a class="anchor" aria-label="anchor" href="#arg-subset-row"></a></dt>
<dd><p>A character, integer or logical vector specifying the genes to use for the velocity calculations.
Defaults to all genes.</p></dd>


<dt id="arg-sf-x">sf.X<a class="anchor" aria-label="anchor" href="#arg-sf-x"></a></dt>
<dd><p>A numeric vector containing size factors for usual count matrix.
Defaults to <code><a href="https://rdrr.io/pkg/scuttle/man/librarySizeFactors.html" class="external-link">librarySizeFactors</a></code> on the <code>"X"</code> matrix in <code>x</code>.</p></dd>


<dt id="arg-sf-spliced">sf.spliced<a class="anchor" aria-label="anchor" href="#arg-sf-spliced"></a></dt>
<dd><p>A numeric vector containing size factors for the spliced counts for each cell.
Defaults to <code><a href="https://rdrr.io/pkg/scuttle/man/librarySizeFactors.html" class="external-link">librarySizeFactors</a></code> on the <code>"spliced"</code> matrix in <code>x</code>.</p></dd>


<dt id="arg-sf-unspliced">sf.unspliced<a class="anchor" aria-label="anchor" href="#arg-sf-unspliced"></a></dt>
<dd><p>A numeric vector containing size factors for the unspliced counts for each cell.
Defaults to <code><a href="https://rdrr.io/pkg/scuttle/man/librarySizeFactors.html" class="external-link">librarySizeFactors</a></code> on the <code>"unspliced"</code> matrix in <code>x</code>.</p></dd>


<dt id="arg-use-theirs">use.theirs<a class="anchor" aria-label="anchor" href="#arg-use-theirs"></a></dt>
<dd><p>Logical scalar indicating whether <span class="pkg">scVelo</span>'s gene filtering and normalization should be used.</p></dd>


<dt id="arg-mode">mode<a class="anchor" aria-label="anchor" href="#arg-mode"></a></dt>
<dd><p>String specifying the method to use to estimate the transcriptional dynamics.</p></dd>


<dt id="arg-scvelo-params">scvelo.params<a class="anchor" aria-label="anchor" href="#arg-scvelo-params"></a></dt>
<dd><p>List of lists containing arguments for individual <span class="pkg">scVelo</span> functions, see details below.</p></dd>


<dt id="arg-dimred">dimred<a class="anchor" aria-label="anchor" href="#arg-dimred"></a></dt>
<dd><p>A low-dimensional representation of the cells with number of rows equal to the number of cells in <code>x</code>,
used to find the nearest neighbors.</p></dd>


<dt id="arg-ncomponents">ncomponents<a class="anchor" aria-label="anchor" href="#arg-ncomponents"></a></dt>
<dd><p>Numeric scalar indicating the number of principal components to obtain.
Only used if <code>use.theirs=FALSE</code> and <code>dimred=NULL</code>.</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>A BiocParallelParam object specifying whether the PCA calculations should be parallelized.
Only used if <code>use.theirs=FALSE</code> and <code>dimred=NULL</code>.</p></dd>


<dt id="arg-bsparam">BSPARAM<a class="anchor" aria-label="anchor" href="#arg-bsparam"></a></dt>
<dd><p>A BiocSingularParam object specifying which algorithm should be used to perform the PCA.
Only used if <code>use.theirs=FALSE</code> and <code>dimred=NULL</code>.</p></dd>


<dt id="arg-assay-x">assay.X<a class="anchor" aria-label="anchor" href="#arg-assay-x"></a></dt>
<dd><p>An integer scalar or string specifying the assay of <code>x</code> containing the usual count matrix.</p></dd>


<dt id="arg-assay-spliced">assay.spliced<a class="anchor" aria-label="anchor" href="#arg-assay-spliced"></a></dt>
<dd><p>An integer scalar or string specifying the assay of <code>x</code> containing the spliced counts.</p></dd>


<dt id="arg-assay-unspliced">assay.unspliced<a class="anchor" aria-label="anchor" href="#arg-assay-unspliced"></a></dt>
<dd><p>An integer scalar or string specifying the assay of <code>x</code> containing the unspliced counts.</p></dd>


<dt id="arg-use-dimred">use.dimred<a class="anchor" aria-label="anchor" href="#arg-use-dimred"></a></dt>
<dd><p>String naming the entry of <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDims</a>(x)</code> to use for nearest neighbor calculations.
Ignored if <code>dimred</code> is supplied.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a> is returned containing the output of the velocity calculations.
Of particular interest are:</p><ul><li><p>the <code>velocity_pseudotime</code> field in the <code>colData</code>,
containing the velocity pseudotime for each cell.</p></li>
<li><p>the <code>velocity</code> entry of the <code>assays</code>,
containing the velocity vectors for each cell.</p></li>
</ul><p>The output will always have number of columns equal to the number of cells supplied in <code>x</code>,
though the number of rows will depend on whether any subsetting (if <code>subset.row</code> is supplied)
or feature selection (if <code>use.theirs=TRUE</code>) was performed.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function uses the <span class="pkg">scVelo</span> Python package (<a href="https://pypi.org/project/scvelo/" class="external-link">https://pypi.org/project/scvelo/</a>) for RNA velocity calculations.
The main difference from the original <span class="pkg">velocyto</span> approach is that the dynamical model of <span class="pkg">scVelo</span>
does not rely on the presence of observed steady-state populations,
which should improve the reliability of the velocity calculations in general applications.</p>
<p>For consistency with other Bioconductor workflows, we perform as many standard steps in R as we can
before starting the velocity calculations with <span class="pkg">scVelo</span>.
This involves:</p><ol><li><p>Size factor-based normalization with <code>sf.*</code> values and <code><a href="https://rdrr.io/pkg/scuttle/man/normalizeCounts.html" class="external-link">normalizeCounts</a></code>.
For <code>"X"</code>, log-transformation is performed as well, while for the others, only scaling normalization is performed.</p></li>
<li><p>Subsetting all matrices to <code>subset.row</code>, most typically to a subset of interest, e.g., highly variable genes.
Note that, if set, any subsetting is done <em>after</em> normalization so that library sizes are correctly computed.</p></li>
<li><p>If <code>dimred=NULL</code>, the PCA step on the log-expression values derived from the <code>"X"</code> matrix,
using the specified <code>BSPARAM</code> to obtain the first <code>ncomponents</code> PCs.</p></li>
</ol><p>This allows us to guarantee that, for example, the log-expression matrix of HVGs or the PCA coordinates
are the same as that used in other applications like clustering or trajectory reconstruction.</p>
<p>Nonetheless, one can set <code>use.theirs=TRUE</code> to directly use the entire <span class="pkg">scVelo</span> normalization and filtering pipeline.
This ignores all of the size factors arguments (<code>sf.*</code>),
all of the PCA-related arguments (<code>ncomponents</code>, <code>BSPARAM</code>) and <code>subset.row</code>.
However, if a low-dimensionality result is supplied via <code>dimred</code> or <code>use.dimred</code>,
the <span class="pkg">scVelo</span> PCA will always be omitted.</p>
<p>Upon first use, this function will instantiate a conda environment containing the <span class="pkg">scVelo</span> package.
This is done via the <span class="pkg">basilisk</span> package - see the documentation for that package for trouble-shooting.</p>
    </div>
    <div id="comments-on-the-three-matrices">
    <h2>Comments on the three matrices</h2>


<p>Strictly speaking, only the spliced and unspliced matrices are necessary for the velocity calculations.
However, it is often the case that the spliced matrix is not actually the same as a “usual” count matrix
(e.g., generated by summing counts across all exons for all mapped genes).
This is due to differences in the handling of ambiguous reads that map across exon-intron boundaries,
or to genomic regions that can be either exonic or intronic depending on the isoform;
the spliced count matrix is more likely to exclude such reads.</p>
<p>We request the usual count matrix as the <code>"X"</code> entry of <code>x</code> to ensure that
the PCA and nearest neighbor detection in <span class="pkg">scVelo</span> are done on the same data
as that used in other steps of the large analysis (e.g., clustering, visualization, trajectory reconstruction).
In practice, if the usual count matrix is not available, one can often achieve satisfactory results
by simply re-using the spliced count matrix as both the <code>"X"</code> and <code>"spliced"</code> entries of <code>x</code>.</p>
<p>Note that if reduced dimensions are supplied in <code>dimred</code>,
any <code>"X"</code> entry is only used to create the AnnData object and is not used in any actual calculations.</p>
    </div>
    <div id="additional-arguments-to-python">
    <h2>Additional arguments to Python</h2>


<p>Additional arguments to <span class="pkg">scVelo</span> functions are provided via <code>scvelo.params</code>.
This is a named list where each entry is named after a function and is itself a named list of arguments for that function.
The following function names are currently recognized:</p><ul><li><p><code>"filter_and_normalize"</code>, for gene selection and normalization.
This is not used unless <code>use.theirs=TRUE</code>.</p></li>
<li><p><code>"moments"</code>, for PCA and nearest neighbor detection.
The PCA is not performed if <code>dimred</code> or <code>use.dimred</code> is already supplied.</p></li>
<li><p><code>"recover_dynamics"</code></p></li>
<li><p><code>"velocity"</code></p></li>
<li><p><code>"velocity_graph"</code></p></li>
<li><p><code>"velocity_pseudotime"</code></p></li>
<li><p><code>"latent_time"</code></p></li>
<li><p><code>"velocity_confidence"</code></p></li>
</ul><p>See the <span class="pkg">scVelo</span> documentation for more details about the available arguments and the examples below for a syntax example.</p>
    </div>
    <div id="supported-operating-systems-and-architectures">
    <h2>Supported operating systems and architectures</h2>


<p><span class="pkg">scVelo</span> dependencies are pinned in a Conda environment to ensure reproducibility.</p>
<p>Differences in packages and versions available from Conda require different environments for different operating systems and architectures.
<span class="pkg">basilisk.utils</span> is used to determine the operating system and architecture of the computer used to run <code>scvelo()</code>, using to the appropriate Conda environment.</p>
<p>As of the latest <span class="pkg">velociraptor</span> update (24 May 2024):</p>
<dl><dt>Linux</dt>
<dd><p><span class="pkg">scVelo</span> v0.2.2 from bioconda is used.
This is NOT the latest version available to date.
Instead, it is the latest version successfully tested in a complete Conda environment.
Note that pinned dependencies were obtained from <code><a href="https://rdrr.io/pkg/zellkonverter/man/AnnData-Environment.html" class="external-link">zellkonverter::AnnDataDependencies()</a></code>.</p></dd>

<dt>MacOS Arm</dt>
<dd><p><span class="pkg">scVelo</span> v0.3.2 from conda-forge is used.
This is the latest version available to date.
Tested on M1.</p></dd>

<dt>Windows</dt>
<dd><p><span class="pkg">scVelo</span> v0.2.5 from bioconda is used.
Later versions of <span class="pkg">scVelo</span> depend on <span class="pkg">jaxlib</span> which is not supported on Windows (<a href="https://github.com/google/jax/issues/438" class="external-link">https://github.com/google/jax/issues/438</a>).
Note that <span class="pkg">matplotlib</span> is pinned to v3.6.3 (<a href="https://github.com/scverse/scanpy/issues/2411" class="external-link">https://github.com/scverse/scanpy/issues/2411</a>),
<span class="pkg">pandas</span> is pinned to v1.5.2 (<a href="https://stackoverflow.com/questions/76234312/importerror-cannot-import-name-is-categorical-from-pandas-api-types" class="external-link">https://stackoverflow.com/questions/76234312/importerror-cannot-import-name-is-categorical-from-pandas-api-types</a>),
and <span class="pkg">numpy</span> is pinned to v1.21.1 (<a href="https://github.com/theislab/scvelo/issues/1109" class="external-link">https://github.com/theislab/scvelo/issues/1109</a>).</p></dd>


</dl></div>
    <div id="references">
    <h2>References</h2>
    <p>Bergen, V., Lange, M., Peidli, S. et al. Generalizing RNA velocity to transient cell states through dynamical modeling. Nat Biotechnol 38, 1408–1414 (2020). <a href="https://doi.org/10.1038/s41587-020-0591-3" class="external-link">https://doi.org/10.1038/s41587-020-0591-3</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Aaron Lun, Charlotte Soneson, Kevin Rue-Albrecht</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Using mock data to demonstrate the process:</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/mockSCE.html" class="external-link">mockSCE</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/mockSCE.html" class="external-link">mockSCE</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">spliced</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">unspliced</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce2</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">scvelo</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">spliced</span>, spliced<span class="op">=</span><span class="va">spliced</span>, unspliced<span class="op">=</span><span class="va">unspliced</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># make scvelo use 10 rather than the default 30 neighbors to compute moments for velocity estimation:</span></span></span>
<span class="r-in"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">scvelo</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">spliced</span>, spliced<span class="op">=</span><span class="va">spliced</span>, unspliced<span class="op">=</span><span class="va">unspliced</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>              scvelo.params<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>neighbors<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>n_neighbors<span class="op">=</span><span class="fl">10L</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Rue-Albrecht, Aaron Lun, Charlotte Soneson, Michael Stadler.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer></div>






  </body></html>

