<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Computing RNA velocity in a Bioconductor framework • velociraptor</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Computing RNA velocity in a Bioconductor framework">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">velociraptor</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.15.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/velociraptor.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="divider">
    </li>
<li>
      <a href="../articles/index.html">More articles...</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kevinrue/velociraptor/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Computing RNA velocity in a Bioconductor
framework</h1>
                        <h4 data-toc-skip class="author">Kevin
Rue-Albrecht</h4>
                                    <h4 data-toc-skip class="author">Aaron
Lun</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:infinite.monkeys.with.keyboards@gmail.com" class="email">infinite.monkeys.with.keyboards@gmail.com</a>
      
                              <h4 data-toc-skip class="author">Charlotte
Soneson</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:charlottesoneson@gmail.com" class="email">charlottesoneson@gmail.com</a>
      
                  
      
      <small class="dont-index">Source: <a href="https://github.com/kevinrue/velociraptor/blob/devel/vignettes/velociraptor.Rmd" class="external-link"><code>vignettes/velociraptor.Rmd</code></a></small>
      <div class="hidden name"><code>velociraptor.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This package provides a lightweight interface between the
Bioconductor <code>SingleCellExperiment</code> data structure and the <a href="https://pypi.org/project/scvelo" class="external-link">scvelo</a> Python package for RNA
velocity calculations. The interface is comparable to that of many other
<code>SingleCellExperiment</code>-compatible functions, allowing users
to plug in RNA velocity calculations into the existing Bioconductor
analysis framework. To demonstrate, we will use a data set from <span class="citation">Hermann et al. (2018)</span>, provided via the <em><a href="https://bioconductor.org/packages/3.20/scRNAseq" class="external-link">scRNAseq</a></em>
package. This data set contains gene-wise estimates of spliced and
unspliced UMI counts for 2,325 mouse spermatogenic cells.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/HermannSpermatogenesisData.html" class="external-link">HermannSpermatogenesisData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 54448 2325 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): spliced unspliced</span></span>
<span><span class="co">## rownames(54448): ENSMUSG00000102693.1 ENSMUSG00000064842.1 ...</span></span>
<span><span class="co">##   ENSMUSG00000064369.1 ENSMUSG00000064372.1</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(2325): CCCATACTCCGAAGAG AATCCAGTCATCTGCC ... ATCCACCCACCACCAG</span></span>
<span><span class="co">##   ATTGGTGGTTACCGAT</span></span>
<span><span class="co">## colData names(1): celltype</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="downsampling-for-demonstration">Downsampling for demonstration<a class="anchor" aria-label="anchor" href="#downsampling-for-demonstration"></a>
</h2>
<p>The full data set requires up to 12 GB of memory for the example
usage presented in this vignette. For demonstration purposes, we
downsample the data set to the first 500 cells. Feel free to skip this
downsampling step if you have access to sufficient memory.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="basic-workflow">Basic workflow<a class="anchor" aria-label="anchor" href="#basic-workflow"></a>
</h2>
<p>We assume that feature selection has already been performed by the
user using any method (see <a href="https://osca.bioconductor.org/feature-selection.html" class="external-link">here</a> for
some suggestions). In this case, we will use the variance of
log-expressions from <em><a href="https://bioconductor.org/packages/3.20/scran" class="external-link">scran</a></em> to
select the top 2000 genes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span>, assay.type<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">top.hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n<span class="op">=</span><span class="fl">2000</span><span class="op">)</span></span></code></pre></div>
<p>We can plug these choices into the <code><a href="../reference/scvelo.html">scvelo()</a></code> function
with our <code>SingleCellExperiment</code> object. By default,
<code><a href="../reference/scvelo.html">scvelo()</a></code> uses the steady-state approach to estimate
velocities, though the stochastic and dynamical models implemented in <a href="https://pypi.org/project/scvelo" class="external-link">scvelo</a> can also be used by
modifying the <code>mode</code> argument.</p>
<p>Note that automatic neighbor calculation is deprecated since
scvelo==0.4.0 and will be removed in a future version. Instead, <em><a href="https://bioconductor.org/packages/3.20/velociraptor" class="external-link">velociraptor</a></em>
computes neighbors with Scanpy (as per <a href="https://github.com/theislab/scvelo/blob/f21651c3b122860d8ae6b5a5173f242ba91c8761/scvelo/preprocessing/moments.py#L66" class="external-link">scVelo
recommendations</a>), and the number of neighbors should be supplied to
<a href="https://scanpy.readthedocs.io/en/stable/api/generated/scanpy.pp.neighbors.html" class="external-link">scanpy.pp.neighbors</a>
as demonstrated below.</p>
<p>In particular, the default number of neighbors was 30 for <a href="https://scvelo.readthedocs.io/en/stable/scvelo.pp.moments.html" class="external-link">scvelo.pp.moments</a>
while it is 15 for <a href="https://scanpy.readthedocs.io/en/stable/api/generated/scanpy.pp.neighbors.html" class="external-link">scanpy.pp.neighbors</a>.
Users should use
<code>scvelo.params=list(neighbors=list(n_neighbors=30L)</code> to
reproduce earlier results.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kevinrue/velociraptor" class="external-link">velociraptor</a></span><span class="op">)</span></span>
<span><span class="va">velo.out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scvelo.html">scvelo</a></span><span class="op">(</span></span>
<span>  <span class="va">sce</span>, subset.row<span class="op">=</span><span class="va">top.hvgs</span>, assay.X<span class="op">=</span><span class="st">"spliced"</span>,</span>
<span>  scvelo.params<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>neighbors<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>n_neighbors<span class="op">=</span><span class="fl">30L</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## computing moments based on connectivities</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added </span></span>
<span><span class="co">##     'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)</span></span>
<span><span class="co">## computing velocities</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added </span></span>
<span><span class="co">##     'velocity', velocity vectors for each individual cell (adata.layers)</span></span>
<span><span class="co">## computing velocity graph (using 1/4 cores)</span></span>
<span><span class="co">## WARNING: Unable to create progress bar. Consider installing `tqdm` as `pip install tqdm` and `ipywidgets` as `pip install ipywidgets`,</span></span>
<span><span class="co">## or disable the progress bar using `show_progress_bar=False`.</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added </span></span>
<span><span class="co">##     'velocity_graph', sparse matrix with cosine correlations (adata.uns)</span></span>
<span><span class="co">## computing terminal states</span></span>
<span><span class="co">##     identified 1 region of root cells and 1 region of end points .</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added</span></span>
<span><span class="co">##     'root_cells', root cells of Markov diffusion process (adata.obs)</span></span>
<span><span class="co">##     'end_points', end points of Markov diffusion process (adata.obs)</span></span>
<span><span class="co">## --&gt; added 'velocity_length' (adata.obs)</span></span>
<span><span class="co">## --&gt; added 'velocity_confidence' (adata.obs)</span></span>
<span><span class="co">## --&gt; added 'velocity_confidence_transition' (adata.obs)</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">velo.out</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 2000 500 </span></span>
<span><span class="co">## metadata(4): neighbors velocity_params velocity_graph</span></span>
<span><span class="co">##   velocity_graph_neg</span></span>
<span><span class="co">## assays(6): X spliced ... Mu velocity</span></span>
<span><span class="co">## rownames(2000): ENSMUSG00000117819.1 ENSMUSG00000081984.3 ...</span></span>
<span><span class="co">##   ENSMUSG00000022965.8 ENSMUSG00000094660.2</span></span>
<span><span class="co">## rowData names(4): velocity_gamma velocity_qreg_ratio velocity_r2</span></span>
<span><span class="co">##   velocity_genes</span></span>
<span><span class="co">## colnames(500): CCCATACTCCGAAGAG AATCCAGTCATCTGCC ... CACCTTGTCGTAGGAG</span></span>
<span><span class="co">##   TTCCCAGAGACTAAGT</span></span>
<span><span class="co">## colData names(7): velocity_self_transition root_cells ...</span></span>
<span><span class="co">##   velocity_confidence velocity_confidence_transition</span></span>
<span><span class="co">## reducedDimNames(1): X_pca</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>In the above call, we use the <code>"spliced"</code> count matrix as
a proxy for the typical exonic count matrix. Technically, the latter is
not required for the velocity estimation, but <a href="https://pypi.org/project/scvelo" class="external-link">scvelo</a> needs to perform a PCA
and nearest neighbors search, and we want to ensure that the neighbors
detected inside the function are consistent with the rest of the
analysis workflow (performed on the exonic counts). There are some
subtle differences between the spliced count matrix and the typical
exonic count matrix - see <code><a href="../reference/scvelo.html">?scvelo</a></code> for some commentary about
this - but the spliced counts are generally a satisfactory replacement
if the latter is not available.</p>
<p>The <code><a href="../reference/scvelo.html">scvelo()</a></code> function produces a
<code>SingleCellExperiment</code> containing all of the outputs of the
calculation in Python. Of particular interest is the
<code>velocity_pseudotime</code> vector that captures the relative
progression of each cell along the biological process driving the
velocity vectors. We can visualize this effect below in a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>-SNE
plot generated by <em><a href="https://bioconductor.org/packages/3.20/scater" class="external-link">scater</a></em> on
the top HVGs.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, subset_row<span class="op">=</span><span class="va">top.hvgs</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">sce</span>, dimred<span class="op">=</span><span class="st">"PCA"</span>, perplexity <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">velocity_pseudotime</span> <span class="op">&lt;-</span> <span class="va">velo.out</span><span class="op">$</span><span class="va">velocity_pseudotime</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by<span class="op">=</span><span class="st">"velocity_pseudotime"</span><span class="op">)</span></span></code></pre></div>
<p><img src="velociraptor_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>It is also straightforward to embed the velocity vectors into our
desired low-dimensional space, as shown below for the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>-SNE
coordinates. This uses a grid-based approach to summarize the per-cell
vectors into local representatives for effective visualization.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">embedded</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/embedVelocity.html">embedVelocity</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"TSNE"</span><span class="op">)</span>, <span class="va">velo.out</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## computing velocity embedding</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added</span></span>
<span><span class="co">##     'velocity_target', embedded velocity vectors (adata.obsm)</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grid.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gridVectors.html">gridVectors</a></span><span class="op">(</span><span class="va">sce</span>, <span class="va">embedded</span>, use.dimred <span class="op">=</span> <span class="st">"TSNE"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by<span class="op">=</span><span class="st">"velocity_pseudotime"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html" class="external-link">geom_segment</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">grid.df</span>, mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">start.1</span>, y<span class="op">=</span><span class="va">start.2</span>, </span>
<span>        xend<span class="op">=</span><span class="va">end.1</span>, yend<span class="op">=</span><span class="va">end.2</span>, colour<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span>, arrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grid/arrow.html" class="external-link">arrow</a></span><span class="op">(</span>length<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="st">"inches"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="velociraptor_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>And that’s it, really.</p>
</div>
<div class="section level2">
<h2 id="advanced-options">Advanced options<a class="anchor" aria-label="anchor" href="#advanced-options"></a>
</h2>
<p><code><a href="../reference/scvelo.html">scvelo()</a></code> interally performs a PCA step that we can
bypass by supplying our own PC coordinates. Indeed, it is often the case
that we have already performed PCA in the earlier analysis steps, so we
can just re-use those results to (i) save time and (ii) improve
consistency with the other steps. Here, we computed the PCA coordinates
in <code><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA()</a></code> above, so let’s just recycle that:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Only setting assay.X= for the initial AnnData creation,</span></span>
<span><span class="co"># it is not actually used in any further steps.</span></span>
<span><span class="va">velo.out2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scvelo.html">scvelo</a></span><span class="op">(</span><span class="va">sce</span>, assay.X<span class="op">=</span><span class="fl">1</span>, subset.row<span class="op">=</span><span class="va">top.hvgs</span>, use.dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## computing moments based on connectivities</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added </span></span>
<span><span class="co">##     'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)</span></span>
<span><span class="co">## computing velocities</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added </span></span>
<span><span class="co">##     'velocity', velocity vectors for each individual cell (adata.layers)</span></span>
<span><span class="co">## computing velocity graph (using 1/4 cores)</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added </span></span>
<span><span class="co">##     'velocity_graph', sparse matrix with cosine correlations (adata.uns)</span></span>
<span><span class="co">## computing terminal states</span></span>
<span><span class="co">##     identified 3 regions of root cells and 1 region of end points .</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added</span></span>
<span><span class="co">##     'root_cells', root cells of Markov diffusion process (adata.obs)</span></span>
<span><span class="co">##     'end_points', end points of Markov diffusion process (adata.obs)</span></span>
<span><span class="co">## --&gt; added 'velocity_length' (adata.obs)</span></span>
<span><span class="co">## --&gt; added 'velocity_confidence' (adata.obs)</span></span>
<span><span class="co">## --&gt; added 'velocity_confidence_transition' (adata.obs)</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">velo.out2</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 2000 500 </span></span>
<span><span class="co">## metadata(4): neighbors velocity_params velocity_graph</span></span>
<span><span class="co">##   velocity_graph_neg</span></span>
<span><span class="co">## assays(6): X spliced ... Mu velocity</span></span>
<span><span class="co">## rownames(2000): ENSMUSG00000117819.1 ENSMUSG00000081984.3 ...</span></span>
<span><span class="co">##   ENSMUSG00000022965.8 ENSMUSG00000094660.2</span></span>
<span><span class="co">## rowData names(4): velocity_gamma velocity_qreg_ratio velocity_r2</span></span>
<span><span class="co">##   velocity_genes</span></span>
<span><span class="co">## colnames(500): CCCATACTCCGAAGAG AATCCAGTCATCTGCC ... CACCTTGTCGTAGGAG</span></span>
<span><span class="co">##   TTCCCAGAGACTAAGT</span></span>
<span><span class="co">## colData names(7): velocity_self_transition root_cells ...</span></span>
<span><span class="co">##   velocity_confidence velocity_confidence_transition</span></span>
<span><span class="co">## reducedDimNames(1): X_pca</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>We also provide an option to use the <a href="https://pypi.org/project/scvelo" class="external-link">scvelo</a> pipeline without
modification, i.e., relying on their normalization and feature
selection. This sacrifices consistency with other Bioconductor workflows
but enables perfect mimicry of a pure Python-based analysis. In this
case, arguments like <code>subset.row=</code> are simply ignored.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">velo.out3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scvelo.html">scvelo</a></span><span class="op">(</span><span class="va">sce</span>, assay.X<span class="op">=</span><span class="fl">1</span>, use.theirs<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.</span></span>
<span><span class="co">## WARNING: Did not normalize spliced as it looks processed already. To enforce normalization, set `enforce=True`.</span></span>
<span><span class="co">## WARNING: Did not normalize unspliced as it looks processed already. To enforce normalization, set `enforce=True`.</span></span>
<span><span class="co">## Logarithmized X.</span></span>
<span><span class="co">## computing moments based on connectivities</span></span>
<span><span class="co">##     finished (0:00:01) --&gt; added </span></span>
<span><span class="co">##     'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)</span></span>
<span><span class="co">## computing velocities</span></span>
<span><span class="co">##     finished (0:00:01) --&gt; added </span></span>
<span><span class="co">##     'velocity', velocity vectors for each individual cell (adata.layers)</span></span>
<span><span class="co">## computing velocity graph (using 1/4 cores)</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added </span></span>
<span><span class="co">##     'velocity_graph', sparse matrix with cosine correlations (adata.uns)</span></span>
<span><span class="co">## computing terminal states</span></span>
<span><span class="co">##     identified 1 region of root cells and 1 region of end points .</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added</span></span>
<span><span class="co">##     'root_cells', root cells of Markov diffusion process (adata.obs)</span></span>
<span><span class="co">##     'end_points', end points of Markov diffusion process (adata.obs)</span></span>
<span><span class="co">## --&gt; added 'velocity_length' (adata.obs)</span></span>
<span><span class="co">## --&gt; added 'velocity_confidence' (adata.obs)</span></span>
<span><span class="co">## --&gt; added 'velocity_confidence_transition' (adata.obs)</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">velo.out3</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 54448 500 </span></span>
<span><span class="co">## metadata(6): log1p pca ... velocity_graph velocity_graph_neg</span></span>
<span><span class="co">## assays(6): X spliced ... Mu velocity</span></span>
<span><span class="co">## rownames(54448): ENSMUSG00000102693.1 ENSMUSG00000064842.1 ...</span></span>
<span><span class="co">##   ENSMUSG00000064369.1 ENSMUSG00000064372.1</span></span>
<span><span class="co">## rowData names(5): velocity_gamma velocity_qreg_ratio velocity_r2</span></span>
<span><span class="co">##   velocity_genes varm</span></span>
<span><span class="co">## colnames(500): CCCATACTCCGAAGAG AATCCAGTCATCTGCC ... CACCTTGTCGTAGGAG</span></span>
<span><span class="co">##   TTCCCAGAGACTAAGT</span></span>
<span><span class="co">## colData names(11): initial_size_unspliced initial_size_spliced ...</span></span>
<span><span class="co">##   velocity_confidence velocity_confidence_transition</span></span>
<span><span class="co">## reducedDimNames(1): X_pca</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>Advanced users can tinker with the settings of individual <a href="https://pypi.org/project/scvelo" class="external-link">scvelo</a> steps by setting named
lists of arguments in the <code>scvelo.params=</code> argument. For
example, to tinker with the behavior of the
<code>recover_dynamics</code> step, we could do:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">velo.out4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scvelo.html">scvelo</a></span><span class="op">(</span><span class="va">sce</span>, assay.X<span class="op">=</span><span class="fl">1</span>, subset.row<span class="op">=</span><span class="va">top.hvgs</span>,</span>
<span>    scvelo.params<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>recover_dynamics<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>max_iter<span class="op">=</span><span class="fl">20</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## computing moments based on connectivities</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added </span></span>
<span><span class="co">##     'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)</span></span>
<span><span class="co">## computing velocities</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added </span></span>
<span><span class="co">##     'velocity', velocity vectors for each individual cell (adata.layers)</span></span>
<span><span class="co">## computing velocity graph (using 1/4 cores)</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added </span></span>
<span><span class="co">##     'velocity_graph', sparse matrix with cosine correlations (adata.uns)</span></span>
<span><span class="co">## computing terminal states</span></span>
<span><span class="co">##     identified 2 regions of root cells and 1 region of end points .</span></span>
<span><span class="co">##     finished (0:00:00) --&gt; added</span></span>
<span><span class="co">##     'root_cells', root cells of Markov diffusion process (adata.obs)</span></span>
<span><span class="co">##     'end_points', end points of Markov diffusion process (adata.obs)</span></span>
<span><span class="co">## --&gt; added 'velocity_length' (adata.obs)</span></span>
<span><span class="co">## --&gt; added 'velocity_confidence' (adata.obs)</span></span>
<span><span class="co">## --&gt; added 'velocity_confidence_transition' (adata.obs)</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">velo.out4</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 2000 500 </span></span>
<span><span class="co">## metadata(4): neighbors velocity_params velocity_graph</span></span>
<span><span class="co">##   velocity_graph_neg</span></span>
<span><span class="co">## assays(6): X spliced ... Mu velocity</span></span>
<span><span class="co">## rownames(2000): ENSMUSG00000117819.1 ENSMUSG00000081984.3 ...</span></span>
<span><span class="co">##   ENSMUSG00000022965.8 ENSMUSG00000094660.2</span></span>
<span><span class="co">## rowData names(4): velocity_gamma velocity_qreg_ratio velocity_r2</span></span>
<span><span class="co">##   velocity_genes</span></span>
<span><span class="co">## colnames(500): CCCATACTCCGAAGAG AATCCAGTCATCTGCC ... CACCTTGTCGTAGGAG</span></span>
<span><span class="co">##   TTCCCAGAGACTAAGT</span></span>
<span><span class="co">## colData names(7): velocity_self_transition root_cells ...</span></span>
<span><span class="co">##   velocity_confidence velocity_confidence_transition</span></span>
<span><span class="co">## reducedDimNames(1): X_pca</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] scater_1.33.4               ggplot2_3.5.1              </span></span>
<span><span class="co">##  [3] velociraptor_1.15.8         scran_1.33.1               </span></span>
<span><span class="co">##  [5] scuttle_1.15.4              scRNAseq_2.19.1            </span></span>
<span><span class="co">##  [7] SingleCellExperiment_1.27.2 SummarizedExperiment_1.35.1</span></span>
<span><span class="co">##  [9] Biobase_2.65.1              GenomicRanges_1.57.1       </span></span>
<span><span class="co">## [11] GenomeInfoDb_1.41.1         IRanges_2.39.2             </span></span>
<span><span class="co">## [13] S4Vectors_0.43.2            BiocGenerics_0.51.0        </span></span>
<span><span class="co">## [15] MatrixGenerics_1.17.0       matrixStats_1.3.0          </span></span>
<span><span class="co">## [17] knitr_1.48                  BiocStyle_2.33.1           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] jsonlite_1.8.8           magrittr_2.0.3           ggbeeswarm_0.7.2        </span></span>
<span><span class="co">##   [4] GenomicFeatures_1.57.0   gypsum_1.1.6             farver_2.1.2            </span></span>
<span><span class="co">##   [7] rmarkdown_2.28           fs_1.6.4                 BiocIO_1.15.2           </span></span>
<span><span class="co">##  [10] zlibbioc_1.51.1          ragg_1.3.2               vctrs_0.6.5             </span></span>
<span><span class="co">##  [13] memoise_2.0.1            Rsamtools_2.21.1         RCurl_1.98-1.16         </span></span>
<span><span class="co">##  [16] htmltools_0.5.8.1        S4Arrays_1.5.7           AnnotationHub_3.13.3    </span></span>
<span><span class="co">##  [19] curl_5.2.2               BiocNeighbors_1.23.0     Rhdf5lib_1.27.0         </span></span>
<span><span class="co">##  [22] SparseArray_1.5.31       rhdf5_2.49.0             sass_0.4.9              </span></span>
<span><span class="co">##  [25] alabaster.base_1.5.6     bslib_0.8.0              htmlwidgets_1.6.4       </span></span>
<span><span class="co">##  [28] basilisk_1.17.2          desc_1.4.3               alabaster.sce_1.5.1     </span></span>
<span><span class="co">##  [31] httr2_1.0.3              cachem_1.1.0             GenomicAlignments_1.41.0</span></span>
<span><span class="co">##  [34] igraph_2.0.3             lifecycle_1.0.4          pkgconfig_2.0.3         </span></span>
<span><span class="co">##  [37] rsvd_1.0.5               Matrix_1.7-0             R6_2.5.1                </span></span>
<span><span class="co">##  [40] fastmap_1.2.0            GenomeInfoDbData_1.2.12  digest_0.6.37           </span></span>
<span><span class="co">##  [43] colorspace_2.1-1         AnnotationDbi_1.67.0     dqrng_0.4.1             </span></span>
<span><span class="co">##  [46] irlba_2.3.5.1            ExperimentHub_2.13.1     textshaping_0.4.0       </span></span>
<span><span class="co">##  [49] RSQLite_2.3.7            beachmat_2.21.5          labeling_0.4.3          </span></span>
<span><span class="co">##  [52] filelock_1.0.3           fansi_1.0.6              httr_1.4.7              </span></span>
<span><span class="co">##  [55] abind_1.4-5              compiler_4.4.1           withr_3.0.1             </span></span>
<span><span class="co">##  [58] bit64_4.0.5              BiocParallel_1.39.0      viridis_0.6.5           </span></span>
<span><span class="co">##  [61] DBI_1.2.3                highr_0.11               HDF5Array_1.33.6        </span></span>
<span><span class="co">##  [64] alabaster.ranges_1.5.2   alabaster.schemas_1.5.0  rappdirs_0.3.3          </span></span>
<span><span class="co">##  [67] DelayedArray_0.31.11     rjson_0.2.22             bluster_1.15.0          </span></span>
<span><span class="co">##  [70] tools_4.4.1              vipor_0.4.7              beeswarm_0.4.0          </span></span>
<span><span class="co">##  [73] glue_1.7.0               restfulr_0.0.15          rhdf5filters_1.17.0     </span></span>
<span><span class="co">##  [76] grid_4.4.1               Rtsne_0.17               cluster_2.1.6           </span></span>
<span><span class="co">##  [79] generics_0.1.3           gtable_0.3.5             ensembldb_2.29.1        </span></span>
<span><span class="co">##  [82] metapod_1.13.0           BiocSingular_1.21.2      ScaledMatrix_1.13.0     </span></span>
<span><span class="co">##  [85] utf8_1.2.4               XVector_0.45.0           ggrepel_0.9.5           </span></span>
<span><span class="co">##  [88] BiocVersion_3.20.0       pillar_1.9.0             limma_3.61.9            </span></span>
<span><span class="co">##  [91] dplyr_1.1.4              BiocFileCache_2.13.0     lattice_0.22-6          </span></span>
<span><span class="co">##  [94] rtracklayer_1.65.0       bit_4.0.5                tidyselect_1.2.1        </span></span>
<span><span class="co">##  [97] locfit_1.5-9.10          Biostrings_2.73.1        gridExtra_2.3           </span></span>
<span><span class="co">## [100] bookdown_0.40            ProtGenerics_1.37.1      edgeR_4.3.13            </span></span>
<span><span class="co">## [103] xfun_0.47                statmod_1.5.0            UCSC.utils_1.1.0        </span></span>
<span><span class="co">## [106] lazyeval_0.2.2           yaml_2.3.10              evaluate_0.24.0         </span></span>
<span><span class="co">## [109] codetools_0.2-20         tibble_3.2.1             alabaster.matrix_1.5.5  </span></span>
<span><span class="co">## [112] BiocManager_1.30.25      cli_3.6.3                reticulate_1.38.0       </span></span>
<span><span class="co">## [115] systemfonts_1.1.0        munsell_0.5.1            jquerylib_0.1.4         </span></span>
<span><span class="co">## [118] zellkonverter_1.15.1     Rcpp_1.0.13              dir.expiry_1.13.0       </span></span>
<span><span class="co">## [121] dbplyr_2.5.0             png_0.1-8                XML_3.99-0.17           </span></span>
<span><span class="co">## [124] parallel_4.4.1           pkgdown_2.1.0            blob_1.2.4              </span></span>
<span><span class="co">## [127] basilisk.utils_1.17.2    AnnotationFilter_1.29.0  bitops_1.0-8            </span></span>
<span><span class="co">## [130] viridisLite_0.4.2        alabaster.se_1.5.3       scales_1.3.0            </span></span>
<span><span class="co">## [133] crayon_1.5.3             rlang_1.1.4              cowplot_1.1.3           </span></span>
<span><span class="co">## [136] KEGGREST_1.45.1</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Hermann2018-spermatogenesis" class="csl-entry">
Hermann, Brian P, Keren Cheng, Anukriti Singh, Lorena Roa-De La Cruz,
Kazadi N Mutoji, I-Chung Chen, Heidi Gildersleeve, et al. 2018.
<span>“The Mammalian Spermatogenesis <span>Single-Cell</span>
Transcriptome, from Spermatogonial Stem Cells to Spermatids.”</span>
<em>Cell Rep.</em> 25: 1650–1667.e8.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Rue-Albrecht, Aaron Lun, Charlotte Soneson, Michael Stadler.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
