<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Computing RNA velocity in a Bioconductor framework • velociraptor</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Computing RNA velocity in a Bioconductor framework">
<meta property="og:description" content="velociraptor">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">velociraptor</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/velociraptor.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="divider">
    </li>
<li>
      <a href="../articles/index.html">More articles...</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kevinrue/velociraptor/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Computing RNA velocity in a Bioconductor framework</h1>
                        <h4 data-toc-skip class="author">Kevin Rue-Albrecht</h4>
                                    <h4 data-toc-skip class="author">Aaron Lun</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:infinite.monkeys.with.keyboards@gmail.com" class="email">infinite.monkeys.with.keyboards@gmail.com</a>
      
                              <h4 data-toc-skip class="author">Charlotte Soneson</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:charlottesoneson@gmail.com" class="email">charlottesoneson@gmail.com</a>
      
                  
      
      <small class="dont-index">Source: <a href="https://github.com/kevinrue/velociraptor/blob/master/vignettes/velociraptor.Rmd" class="external-link"><code>vignettes/velociraptor.Rmd</code></a></small>
      <div class="hidden name"><code>velociraptor.Rmd</code></div>

    </div>

    
    
<div class="section level1">
<h1 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h1>
<p>This package provides a lightweight interface between the Bioconductor <code>SingleCellExperiment</code> data structure and the <a href="https://pypi.org/project/scvelo" class="external-link">scvelo</a> Python package for RNA velocity calculations. The interface is comparable to that of many other <code>SingleCellExperiment</code>-compatible functions, allowing users to plug in RNA velocity calculations into the existing Bioconductor analysis framework. To demonstrate, we will use a data set from <span class="citation">Hermann et al. (2018)</span>, provided via the <em><a href="https://bioconductor.org/packages/3.14/scRNAseq" class="external-link">scRNAseq</a></em> package. This data set contains gene-wise estimates of spliced and unspliced UMI counts for 2,325 mouse spermatogenic cells.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/HermannSpermatogenesisData.html" class="external-link">HermannSpermatogenesisData</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">sce</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 54448 2325 
## metadata(0):
## assays(2): spliced unspliced
## rownames(54448): ENSMUSG00000102693.1 ENSMUSG00000064842.1 ...
##   ENSMUSG00000064369.1 ENSMUSG00000064372.1
## rowData names(0):
## colnames(2325): CCCATACTCCGAAGAG AATCCAGTCATCTGCC ... ATCCACCCACCACCAG
##   ATTGGTGGTTACCGAT
## colData names(1): celltype
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):</code></pre>
</div>
<div class="section level1">
<h1 id="downsampling-for-demonstration">Downsampling for demonstration<a class="anchor" aria-label="anchor" href="#downsampling-for-demonstration"></a>
</h1>
<p>The full data set requires up to 12 GB of memory for the example usage presented in this vignette. For demonstration purposes, we downsample the data set to the first 500 cells. Feel free to skip this downsampling step if you have access to sufficient memory.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="basic-workflow">Basic workflow<a class="anchor" aria-label="anchor" href="#basic-workflow"></a>
</h1>
<p>We assume that feature selection has already been performed by the user using any method (see <a href="https://osca.bioconductor.org/feature-selection.html" class="external-link">here</a> for some suggestions). In this case, we will use the variance of log-expressions from <em><a href="https://bioconductor.org/packages/3.14/scran" class="external-link">scran</a></em> to select the top 2000 genes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span>, assay.type<span class="op">=</span><span class="fl">1</span><span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span>
<span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="va">top.hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n<span class="op">=</span><span class="fl">2000</span><span class="op">)</span></code></pre></div>
<p>We can plug these choices into the <code><a href="../reference/scvelo.html">scvelo()</a></code> function with our <code>SingleCellExperiment</code> object. By default, <code><a href="../reference/scvelo.html">scvelo()</a></code> uses the steady-state approach to estimate velocities, though the stochastic and dynamical models implemented in <a href="https://pypi.org/project/scvelo" class="external-link">scvelo</a> can also be used by modifying the <code>mode</code> argument.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kevinrue/velociraptor" class="external-link">velociraptor</a></span><span class="op">)</span>
<span class="va">velo.out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scvelo.html">scvelo</a></span><span class="op">(</span><span class="va">sce</span>, subset.row<span class="op">=</span><span class="va">top.hvgs</span>, assay.X<span class="op">=</span><span class="st">"spliced"</span><span class="op">)</span>
<span class="va">velo.out</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 2000 500 
## metadata(4): neighbors velocity_params velocity_graph
##   velocity_graph_neg
## assays(6): X spliced ... Mu velocity
## rownames(2000): ENSMUSG00000117819.1 ENSMUSG00000081984.3 ...
##   ENSMUSG00000022965.8 ENSMUSG00000094660.2
## rowData names(3): velocity_gamma velocity_r2 velocity_genes
## colnames(500): CCCATACTCCGAAGAG AATCCAGTCATCTGCC ... CACCTTGTCGTAGGAG
##   TTCCCAGAGACTAAGT
## colData names(7): velocity_self_transition root_cells ...
##   velocity_confidence velocity_confidence_transition
## reducedDimNames(1): X_pca
## mainExpName: NULL
## altExpNames(0):</code></pre>
<p>In the above call, we use the <code>"spliced"</code> count matrix as a proxy for the typical exonic count matrix. Technically, the latter is not required for the velocity estimation, but <a href="https://pypi.org/project/scvelo" class="external-link">scvelo</a> needs to perform a PCA and nearest neighbors search, and we want to ensure that the neighbors detected inside the function are consistent with the rest of the analysis workflow (performed on the exonic counts). There are some subtle differences between the spliced count matrix and the typical exonic count matrix - see <code><a href="../reference/scvelo.html">?scvelo</a></code> for some commentary about this - but the spliced counts are generally a satisfactory replacement if the latter is not available.</p>
<p>The <code><a href="../reference/scvelo.html">scvelo()</a></code> function produces a <code>SingleCellExperiment</code> containing all of the outputs of the calculation in Python. Of particular interest is the <code>velocity_pseudotime</code> vector that captures the relative progression of each cell along the biological process driving the velocity vectors. We can visualize this effect below in a <span class="math inline">\(t\)</span>-SNE plot generated by <em><a href="https://bioconductor.org/packages/3.14/scater" class="external-link">scater</a></em> on the top HVGs.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, subset_row<span class="op">=</span><span class="va">top.hvgs</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">sce</span>, dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span>

<span class="va">sce</span><span class="op">$</span><span class="va">velocity_pseudotime</span> <span class="op">&lt;-</span> <span class="va">velo.out</span><span class="op">$</span><span class="va">velocity_pseudotime</span>
<span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by<span class="op">=</span><span class="st">"velocity_pseudotime"</span><span class="op">)</span></code></pre></div>
<p><img src="velociraptor_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>It is also straightforward to embed the velocity vectors into our desired low-dimensional space, as shown below for the <span class="math inline">\(t\)</span>-SNE coordinates. This uses a grid-based approach to summarize the per-cell vectors into local representatives for effective visualization.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">embedded</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/embedVelocity.html">embedVelocity</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"TSNE"</span><span class="op">)</span>, <span class="va">velo.out</span><span class="op">)</span>
<span class="va">grid.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gridVectors.html">gridVectors</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"TSNE"</span><span class="op">)</span>, <span class="va">embedded</span><span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by<span class="op">=</span><span class="st">"velocity_pseudotime"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html" class="external-link">geom_segment</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">grid.df</span>, mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">start.1</span>, y<span class="op">=</span><span class="va">start.2</span>, 
        xend<span class="op">=</span><span class="va">end.1</span>, yend<span class="op">=</span><span class="va">end.2</span><span class="op">)</span>, arrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grid/arrow.html" class="external-link">arrow</a></span><span class="op">(</span>length<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="st">"inches"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="velociraptor_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>And that’s it, really.</p>
</div>
<div class="section level1">
<h1 id="advanced-options">Advanced options<a class="anchor" aria-label="anchor" href="#advanced-options"></a>
</h1>
<p><code><a href="../reference/scvelo.html">scvelo()</a></code> interally performs a PCA step that we can bypass by supplying our own PC coordinates. Indeed, it is often the case that we have already performed PCA in the earlier analysis steps, so we can just re-use those results to (i) save time and (ii) improve consistency with the other steps. Here, we computed the PCA coordinates in <code><a href="https://rdrr.io/pkg/scater/man/runPCA.html" class="external-link">runPCA()</a></code> above, so let’s just recycle that:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Only setting assay.X= for the initial AnnData creation,</span>
<span class="co"># it is not actually used in any further steps.</span>
<span class="va">velo.out2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scvelo.html">scvelo</a></span><span class="op">(</span><span class="va">sce</span>, assay.X<span class="op">=</span><span class="fl">1</span>, subset.row<span class="op">=</span><span class="va">top.hvgs</span>, use.dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span> 
<span class="va">velo.out2</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 2000 500 
## metadata(4): neighbors velocity_params velocity_graph
##   velocity_graph_neg
## assays(6): X spliced ... Mu velocity
## rownames(2000): ENSMUSG00000117819.1 ENSMUSG00000081984.3 ...
##   ENSMUSG00000022965.8 ENSMUSG00000094660.2
## rowData names(3): velocity_gamma velocity_r2 velocity_genes
## colnames(500): CCCATACTCCGAAGAG AATCCAGTCATCTGCC ... CACCTTGTCGTAGGAG
##   TTCCCAGAGACTAAGT
## colData names(7): velocity_self_transition root_cells ...
##   velocity_confidence velocity_confidence_transition
## reducedDimNames(1): X_pca
## mainExpName: NULL
## altExpNames(0):</code></pre>
<p>We also provide an option to use the <a href="https://pypi.org/project/scvelo" class="external-link">scvelo</a> pipeline without modification, i.e., relying on their normalization and feature selection. This sacrifices consistency with other Bioconductor workflows but enables perfect mimicry of a pure Python-based analysis. In this case, arguments like <code>subset.row=</code> are simply ignored.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">velo.out3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scvelo.html">scvelo</a></span><span class="op">(</span><span class="va">sce</span>, assay.X<span class="op">=</span><span class="fl">1</span>, use.theirs<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">velo.out3</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 54448 500 
## metadata(5): pca neighbors velocity_params velocity_graph
##   velocity_graph_neg
## assays(6): X spliced ... Mu velocity
## rownames(54448): ENSMUSG00000102693.1 ENSMUSG00000064842.1 ...
##   ENSMUSG00000064369.1 ENSMUSG00000064372.1
## rowData names(4): velocity_gamma velocity_r2 velocity_genes varm
## colnames(500): CCCATACTCCGAAGAG AATCCAGTCATCTGCC ... CACCTTGTCGTAGGAG
##   TTCCCAGAGACTAAGT
## colData names(11): initial_size_spliced initial_size_unspliced ...
##   velocity_confidence velocity_confidence_transition
## reducedDimNames(1): X_pca
## mainExpName: NULL
## altExpNames(0):</code></pre>
<p>Advanced users can tinker with the settings of individual <a href="https://pypi.org/project/scvelo" class="external-link">scvelo</a> steps by setting named lists of arguments in the <code>scvelo.params=</code> argument. For example, to tinker with the behavior of the <code>recover_dynamics</code> step, we could do:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">velo.out4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scvelo.html">scvelo</a></span><span class="op">(</span><span class="va">sce</span>, assay.X<span class="op">=</span><span class="fl">1</span>, subset.row<span class="op">=</span><span class="va">top.hvgs</span>,
    scvelo.params<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>recover_dynamics<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>max_iter<span class="op">=</span><span class="fl">20</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">velo.out4</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 2000 500 
## metadata(4): neighbors velocity_params velocity_graph
##   velocity_graph_neg
## assays(6): X spliced ... Mu velocity
## rownames(2000): ENSMUSG00000117819.1 ENSMUSG00000081984.3 ...
##   ENSMUSG00000022965.8 ENSMUSG00000094660.2
## rowData names(3): velocity_gamma velocity_r2 velocity_genes
## colnames(500): CCCATACTCCGAAGAG AATCCAGTCATCTGCC ... CACCTTGTCGTAGGAG
##   TTCCCAGAGACTAAGT
## colData names(7): velocity_self_transition root_cells ...
##   velocity_confidence velocity_confidence_transition
## reducedDimNames(1): X_pca
## mainExpName: NULL
## altExpNames(0):</code></pre>
</div>
<div class="section level1">
<h1 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h1>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.1.1 (2021-08-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.3 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats    graphics utils    stats4   methods  base    
## 
## other attached packages:
##  [1] scater_1.21.8               ggplot2_3.3.5              
##  [3] velociraptor_1.3.1          scran_1.21.3               
##  [5] scuttle_1.3.1               scRNAseq_2.7.2             
##  [7] SingleCellExperiment_1.15.2 SummarizedExperiment_1.23.5
##  [9] Biobase_2.53.0              GenomicRanges_1.45.0       
## [11] GenomeInfoDb_1.29.10        IRanges_2.27.2             
## [13] S4Vectors_0.31.5            BiocGenerics_0.39.2        
## [15] MatrixGenerics_1.5.4        matrixStats_0.61.0         
## [17] knitr_1.36                  BiocStyle_2.21.4           
## 
## loaded via a namespace (and not attached):
##   [1] AnnotationHub_3.1.7           BiocFileCache_2.1.1          
##   [3] systemfonts_1.0.3             igraph_1.2.7                 
##   [5] lazyeval_0.2.2                BiocParallel_1.27.17         
##   [7] digest_0.6.28                 ensembldb_2.17.4             
##   [9] htmltools_0.5.2               viridis_0.6.2                
##  [11] fansi_0.5.0                   magrittr_2.0.1               
##  [13] memoise_2.0.0                 ScaledMatrix_1.1.0           
##  [15] cluster_2.1.2                 limma_3.49.4                 
##  [17] Biostrings_2.61.2             pkgdown_1.9000.9000.9000     
##  [19] prettyunits_1.1.1             colorspace_2.0-2             
##  [21] ggrepel_0.9.1                 blob_1.2.2                   
##  [23] rappdirs_0.3.3                textshaping_0.3.6            
##  [25] xfun_0.27                     dplyr_1.0.7                  
##  [27] crayon_1.4.1                  RCurl_1.98-1.5               
##  [29] jsonlite_1.7.2                glue_1.4.2                   
##  [31] gtable_0.3.0                  zlibbioc_1.39.0              
##  [33] XVector_0.33.0                DelayedArray_0.19.4          
##  [35] BiocSingular_1.9.1            scales_1.1.1                 
##  [37] DBI_1.1.1                     edgeR_3.35.3                 
##  [39] Rcpp_1.0.7                    viridisLite_0.4.0            
##  [41] xtable_1.8-4                  progress_1.2.2               
##  [43] reticulate_1.22               dqrng_0.3.0                  
##  [45] bit_4.0.4                     rsvd_1.0.5                   
##  [47] metapod_1.1.0                 httr_1.4.2                   
##  [49] dir.expiry_1.1.0              ellipsis_0.3.2               
##  [51] farver_2.1.0                  pkgconfig_2.0.3              
##  [53] XML_3.99-0.8                  sass_0.4.0                   
##  [55] dbplyr_2.1.1                  here_1.0.1                   
##  [57] locfit_1.5-9.4                utf8_1.2.2                   
##  [59] labeling_0.4.2                tidyselect_1.1.1             
##  [61] rlang_0.99.0.9000             later_1.3.0                  
##  [63] AnnotationDbi_1.55.2          munsell_0.5.0                
##  [65] BiocVersion_3.14.0            tools_4.1.1                  
##  [67] cachem_1.0.6                  cli_3.0.1                    
##  [69] generics_0.1.0                RSQLite_2.2.8                
##  [71] ExperimentHub_2.1.4           evaluate_0.14                
##  [73] stringr_1.4.0                 fastmap_1.1.0                
##  [75] yaml_2.2.1                    ragg_1.1.3                   
##  [77] grDevices_4.1.1               bit64_4.0.5                  
##  [79] fs_1.5.0                      purrr_0.3.4                  
##  [81] KEGGREST_1.33.0               AnnotationFilter_1.17.1      
##  [83] sparseMatrixStats_1.5.3       mime_0.12                    
##  [85] xml2_1.3.2                    biomaRt_2.49.7               
##  [87] compiler_4.1.1                beeswarm_0.4.0               
##  [89] filelock_1.0.2                curl_4.3.2                   
##  [91] png_0.1-7                     interactiveDisplayBase_1.31.2
##  [93] tibble_3.1.5                  statmod_1.4.36               
##  [95] bslib_0.3.1                   stringi_1.7.5                
##  [97] highr_0.9                     basilisk.utils_1.5.0         
##  [99] GenomicFeatures_1.45.2        desc_1.4.0                   
## [101] lattice_0.20-45               bluster_1.3.2                
## [103] ProtGenerics_1.25.1           Matrix_1.3-4                 
## [105] vctrs_0.3.8                   pillar_1.6.4                 
## [107] lifecycle_1.0.1               BiocManager_1.30.16          
## [109] jquerylib_0.1.4               BiocNeighbors_1.11.0         
## [111] cowplot_1.1.1                 bitops_1.0-7                 
## [113] irlba_2.3.3                   httpuv_1.6.3                 
## [115] rtracklayer_1.53.1            R6_2.5.1                     
## [117] BiocIO_1.3.0                  bookdown_0.24                
## [119] promises_1.2.0.1              gridExtra_2.3                
## [121] vipor_0.4.5                   zellkonverter_1.3.3          
## [123] assertthat_0.2.1              rprojroot_2.0.2              
## [125] rjson_0.2.20                  withr_2.4.2                  
## [127] GenomicAlignments_1.29.0      Rsamtools_2.9.1              
## [129] datasets_4.1.1                GenomeInfoDbData_1.2.7       
## [131] parallel_4.1.1                hms_1.1.1                    
## [133] grid_4.1.1                    beachmat_2.9.1               
## [135] basilisk_1.5.0                rmarkdown_2.11               
## [137] DelayedMatrixStats_1.15.4     Rtsne_0.15                   
## [139] shiny_1.7.1                   ggbeeswarm_0.6.0             
## [141] restfulr_0.0.13</code></pre>
</div>
<div class="section level1 unnumbered">
<h1 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Hermann2018-spermatogenesis" class="csl-entry">
Hermann, Brian P, Keren Cheng, Anukriti Singh, Lorena Roa-De La Cruz, Kazadi N Mutoji, I-Chung Chen, Heidi Gildersleeve, et al. 2018. <span>“The Mammalian Spermatogenesis <span>Single-Cell</span> Transcriptome, from Spermatogonial Stem Cells to Spermatids.”</span> <em>Cell Rep.</em> 25: 1650–1667.e8.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Rue-Albrecht, Aaron Lun, Charlotte Soneson, Michael Stadler.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.9000.9000.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
