#' RNA velocity calculations with \pkg{scVelo}
#'
#' Perform RNA velocity calculations with the \pkg{scVelo} package.
#'
#' @param x A list of three matrices of the same dimensions where genes are in rows and cells are in columns.
#' The first matrix should contain the counts used for calculating reduced dimension representations and cell-to-cell distances.
#' The second matrix should contain spliced counts and the third matrix should contain unspliced counts.
#'
#' Alternatively, a \linkS4class{SummarizedExperiment} object containing three such matrices among its assays.
#' @param ... For the generic, further arguments to pass to specific methods.
#' For the SummarizedExperiment and SingleCellExperiment methods, further arguments to pass to the ANY method.
#' @param assay.x An integer scalar or string specifying the assay of \code{x} containing the counts that will be used for dimension reduction and distance calculations.
#' @param assay.spliced An integer scalar or string specifying the assay of \code{x} containing the spliced counts.
#' @param assay.unspliced An integer scalar or string specifying the assay of \code{x} containing the unspliced counts.
#' @param sf.x A numeric vector containing size factors for the counts in \code{x} for each cell.
#' Defaults to \code{\link{librarySizeFactors}} on the \code{x} count matrix.
#' @param sf.spliced A numeric vector containing size factors for the spliced counts for each cell.
#' Defaults to \code{\link{librarySizeFactors}} on the spliced matrix.
#' @param sf.unspliced A numeric vector containing size factors for the unspliced counts for each cell.
#' Defaults to \code{\link{librarySizeFactors}} on the unspliced matrix.
#' @param subset.row A character, integer or logical vector specifying the genes to use for the velocity calculations.
#' Defaults to all genes but is most typically set to a subset of interesting genes, e.g., highly variable genes.
#' Note that, if set, any subsetting is done \emph{after} normalization so that library sizes are correctly computed.
#' @param use.theirs Logical scalar indicating whether \pkg{scVelo}'s gene filtering and normalization should be used.
#' @param mode String specifying the method to use to estimate the transcriptional dynamics.
#' @param scvelo.params List of lists, providing arguments for scVelo functions.
#' @param dimred A low-dimensional representation of the cells with number of rows equal to the number of cells in \code{x},
#' used to find the nearest neighbors.
#' @param use.dimred String naming a reduced dimension representation to use for nearest neighbor calculations.
#' @param ncomponents Numeric scalar indicating the number of principal components to obtain, 
#' if \code{use.theirs} is FALSE and no \code{dimred} is supplied.
#' @param BSPARAM A \linkS4class{BiocSingularParam} object specifying which algorithm should be used to perform the PCA.
#' @param BPPARAM A \linkS4class{BiocParallelParam} object specifying whether the PCA calculations should be parallelized.
#'
#' @details
#' This function uses the \pkg{scVelo} Python package (\url{https://pypi.org/project/scvelo/}) for RNA velocity calculations.
#' The main difference from the original \pkg{velocyto} approach is that the dynamical model of \pkg{scVelo} 
#' does not rely on the presence of observed steady state populations,
#' which may improve the reliability of the velocity calculations in general applications.
#'
#' For consistency with other Bioconductor workflows, we perform scaling normalization and subsetting in R
#' before starting the velocity calculations with \pkg{scVelo}.
#' This allows us to guarantee that, for example, the log-expression matrix of HVGs or the PCA coordinates
#' are the same as that used in other applications like clustering or trajectory reconstruction.
#' 
#' Nonetheless, one can set \code{use.theirs=TRUE} to directly use the entire \pkg{scVelo} normalization and filtering pipeline.
#' This ignores all of the size factors arguments (\code{sf.*}),
#' all of the PCA-related arguments (\code{ncomponents}, \code{BSPARAM}) and \code{subset.row}.
#' However, if a low-dimensionality result is supplied via \code{dimred} or \code{use.dimred},
#' the \pkg{scVelo} PCA will always be omitted.
#'
#' Upon first use, this function will instantiate a conda environment containing the \pkg{scVelo} package.
#' This is done via the \pkg{basilisk} package - see the documentation for that package for trouble-shooting.
#'
#' @section Comments on the three matrices:
#' It is worth commenting on the use of three separate matrices.
#' Strictly speaking, only the spliced and unspliced matrices are necessary for the velocity calculations.
#' However, it is often the case that the spliced matrix is not actually the same as a \dQuote{usual} count matrix
#' (e.g., generated by summing counts across all exons for all mapped genes).
#' This is due to differences in how ambiguous reads mapping across exon-intron boundaries are handled -
#' quantification of spliced/unspliced counts tends to be more conservative.
#' 
#' We request the usual count matrix as the first entry of \code{x} to ensure that 
#' the PCA and nearest neighbor detection in \pkg{scVelo} are done on the same data 
#' as that used in other steps of the large analysis (e.g., clustering, visualization, trajectory reconstruction).
#' In practice, if the usual count matrix is not available, one can often achieve satisfactory results 
#' by simply re-using the spliced count matrix as both the first and second entries of \code{x}.
#'
#' @section Additional arguments to Python:
#' Additional arguments to \pkg{scVelo} functions are provided via \code{scvelo.params}.
#' This is a named list where each entry is named after a function and is itself a named list of arguments for that function.
#' The following function names are currently recognized:
#' \itemize{
#' \item \code{"filter_and_normalize"}, for gene selection and normalization.
#' This is not used unless \code{use.theirs=TRUE}.
#' \item \code{"moments"}, for PCA and nearest neighbor detection.
#' The PCA is not performed if \code{dimred} or \code{use.dimred} is already supplied.
#' \item \code{"recover_dynamics"}
#' \item \code{"velocity"}
#' \item \code{"velocity_graph"}
#' \item \code{"velocity_pseudotime"}
#' \item \code{"latent_time"}
#' \item \code{"velocity_confidence"}
#' }
#' See the \pkg{scVelo} documentation for more details about the available arguments.
#'
#' @return
#' A \linkS4class{SummarizedExperiment} is returned containing the output of the velocity calculations.
#' TODO: SOME DOCUMENTATION REQUIRED.
#'
#' @examples
#' # Using mock data to demonstrate the process:
#' library(scuttle)
#' sce1 <- mockSCE()
#' sce2 <- mockSCE()
#'
#' spliced <- counts(sce1)
#' unspliced <- counts(sce2)
#'
#' out <- scvelo(list(spliced, spliced, unspliced), use.theirs=TRUE)
#'
#' @references
#' Bergen V. et al. (2019).
#' Generalizing RNA velocity to transient cell states through dynamical modeling.
#' bioxRiv, \url{https://www.biorxiv.org/content/10.1101/820936v1}
#'
#' @author Aaron Lun, Charlotte Soneson
#' @name scvelo
NULL

#' @importFrom S4Vectors DataFrame
#' @importFrom scuttle normalizeCounts
#' @importFrom BiocSingular bsparam runPCA
#' @importFrom BiocParallel SerialParam
#' @importFrom Matrix t
.scvelo <- function(x, subset.row=NULL, 
    sf.x=NULL, sf.spliced=NULL, sf.unspliced=NULL,
    use.theirs=FALSE,
    mode=c('steady_state', 'deterministic', 'stochastic', 'dynamical'),
    scvelo.params=list(), dimred=NULL, ncomponents=30, BPPARAM=SerialParam(),
    BSPARAM=bsparam())
{
    spliced <- x[[2]]
    unspliced <- x[[3]]
    x <- x[[1]]
    if (!all(identical(as.integer(dim(spliced)), as.integer(dim(unspliced))),
             identical(as.integer(dim(spliced)), as.integer(dim(x))))) {
        stop("matrices in 'x' must have the same dimensions")
    }

    if (!use.theirs) {
        x <- normalizeCounts(x, sf.x, log=TRUE)
        spliced <- normalizeCounts(spliced, sf.spliced, log=FALSE)
        unspliced <- normalizeCounts(unspliced, sf.unspliced, log=FALSE)

        if (!is.null(subset.row)) {
            x <- x[subset.row,,drop=FALSE]
            spliced <- spliced[subset.row,,drop=FALSE]
            unspliced <- unspliced[subset.row,,drop=FALSE]
        }

        if (is.null(dimred)) {
            dimred <- BiocSingular::runPCA(t(x), rank=ncomponents, BSPARAM=BSPARAM, BPPARAM=BPPARAM)
        }
    }

    mode <- match.arg(mode)
    output <- basiliskRun(env=velo.env, fun=.run_scvelo,
        x=x, spliced=spliced, unspliced=unspliced,
        use.theirs=use.theirs, mode=mode,
        scvelo.params=scvelo.params,
        dimred=dimred)

    output
}

#' @importFrom reticulate import
#' @importFrom DelayedArray is_sparse t
.run_scvelo <- function(x, spliced, unspliced, use.theirs=FALSE, mode='dynamical', scvelo.params=list(), dimred=NULL) {
    x <- t(.make_np_friendly(x))
    spliced <- t(.make_np_friendly(spliced))
    unspliced <- t(.make_np_friendly(unspliced))

    and <- import("anndata")
    scv <- import("scvelo")
    adata <- and$AnnData(x, layers=list(spliced=spliced, unspliced=unspliced))
    adata$obs_names <- rownames(spliced)
    adata$var_names <- colnames(spliced)

    ## A supplied dimred will be used even if use.theirs=TRUE
    if (!is.null(dimred)) {
        dimred <- .make_np_friendly(dimred)
        adata$obsm <- list(X_pca = dimred)
    }

    if (use.theirs) {
        do.call(scv$pp$filter_and_normalize, c(list(data=adata), scvelo.params$filter_and_normalize))
    }

    do.call(scv$pp$moments, c(list(data=adata), scvelo.params$moments))

    if (mode=="dynamical") {
        do.call(scv$tl$recover_dynamics, c(list(data=adata), scvelo.params$recover_dynamics))
    }

    scvelo.params$velocity$mode <- mode
    do.call(scv$tl$velocity, c(list(data=adata), scvelo.params$velocity))

    do.call(scv$tl$velocity_graph, c(list(data=adata), scvelo.params$velocity_graph))

    do.call(scv$tl$velocity_pseudotime, c(list(adata=adata), scvelo.params$velocity_pseudotime))

    if (mode=="dynamical") {
        do.call(scv$tl$latent_time, c(list(data=adata), scvelo.params$latent_time))
    }

    do.call(scv$tl$velocity_confidence, c(list(data=adata), scvelo.params$velocity_confidence))

    .scvelo_anndata2sce(adata)
}

#' @importFrom S4Vectors make_zero_col_DFrame
.scvelo_anndata2sce <- function(adata) {
    assays <- .extractor_python_dict(adata$layers, c("Mu", "Ms", "velocity"), single=TRUE, transpose=TRUE)
    coldata <- .extractor_python_dict(adata$obs, names(adata$obs))
    rowdata <- .extractor_python_dict(adata$var, names(adata$var))
    metadata <- .extractor_python_dict(adata$uns, c('velocity_params', 'velocity_graph', 'velocity_graph_neg'), single=TRUE)

    if (!is.null(rowdata)) {
        rowdata <- do.call(DataFrame, rowdata)
    } else {
        rowdata <- make_zero_col_DFrame(nrow(assays[[1]]))
        rownames(rowdata) <- rownames(assays[[1]])
    }

    if (!is.null(coldata)) {
        coldata <- do.call(DataFrame, coldata)
    } else {
        coldata <- make_zero_col_DFrame(ncol(assays[[1]]))
        rownames(coldata) <- colnames(assays[[1]])
    }

    SummarizedExperiment(assays, colData=coldata, rowData=rowdata,
        metadata=metadata)
}

#' @export
#' @rdname scvelo
setGeneric("scvelo", function(x, ...) standardGeneric("scvelo"))

#' @export
#' @rdname scvelo
setMethod("scvelo", "ANY", .scvelo)

#' @export
#' @rdname scvelo
#' @importFrom BiocGenerics sizeFactors
setMethod("scvelo", "SummarizedExperiment", function(x, ...,
    assay.x="counts", assay.spliced="spliced", assay.unspliced="unspliced",
    sf.x=sizeFactors(x), sf.spliced=NULL, sf.unspliced=NULL)
{
    .scvelo(list(assay(x, assay.x), assay(x, assay.spliced), assay(x, assay.unspliced)), 
        ..., sf.x=sf.x, sf.spliced=sf.spliced, sf.unspliced=sf.unspliced)
})

#' @export
#' @rdname scvelo
#' @importFrom BiocGenerics sizeFactors
#' @importFrom SingleCellExperiment reducedDimNames reducedDim
setMethod("scvelo", "SingleCellExperiment", function(x, ...,
    assay.x="counts", assay.spliced="spliced", assay.unspliced="unspliced",
    sf.x=sizeFactors(x), sf.spliced=NULL, sf.unspliced=NULL, use.dimred=NULL) 
{
    if (!is.null(use.dimred)) {
        dimred <- reducedDim(x, use.dimred)
    } else {
        dimred <- NULL
    }

    .scvelo(list(assay(x, assay.x), assay(x, assay.spliced), assay(x, assay.unspliced)), 
           ..., sf.x=sf.x, sf.spliced=sf.spliced, sf.unspliced=sf.unspliced, dimred=dimred)
})
